File: examples/ts/oracle-price/OraclePriceFeed.tsop.ts
import { SmartContract, assert, PubKey, Sig, ByteString, RabinSig, RabinPubKey, checkSig, verifyRabinSig, num2bin } from 'tsop-lang';

class OraclePriceFeed extends SmartContract {
  readonly oraclePubKey: RabinPubKey;
  readonly receiver: PubKey;

  constructor(oraclePubKey: RabinPubKey, receiver: PubKey) {
    super(oraclePubKey, receiver);
    this.oraclePubKey = oraclePubKey;
    this.receiver = receiver;
  }

  public settle(price: bigint, rabinSig: RabinSig, padding: ByteString, sig: Sig) {
    // Verify oracle signed this price
    const msg = num2bin(price, 8n);
    assert(verifyRabinSig(msg, rabinSig, padding, this.oraclePubKey));

    // Price must be above threshold for payout
    assert(price > 50000n);

    // Receiver must sign
    assert(checkSig(sig, this.receiver));
  }
}

File: examples/go/oracle-price/OraclePriceFeed.tsop.go
package contract

import "tsop"

type OraclePriceFeed struct {
	tsop.SmartContract
	OraclePubKey tsop.RabinPubKey `tsop:"readonly"`
	Receiver     tsop.PubKey      `tsop:"readonly"`
}

func (c *OraclePriceFeed) Settle(price tsop.Bigint, rabinSig tsop.RabinSig, padding tsop.ByteString, sig tsop.Sig) {
	// Verify oracle signed this price
	msg := tsop.Num2Bin(price, 8)

	tsop.Assert(tsop.VerifyRabinSig(msg, rabinSig, padding, c.OraclePubKey))

	// Price must be above threshold for payout
	tsop.Assert(price > 50000)

	// Receiver must sign
	tsop.Assert(tsop.CheckSig(sig, c.Receiver))
}
